src/sxg_cert_chain.c

```diff
@@ -0,0 +1,336 @@
+
+#include "libsxg/internal/sxg_buffer.h"
+#include "libsxg/internal/sxg_header.h"
+#include "libsxg/sxg_buffer.h"
+
+sxg_cert_t sxg_empty_cert() {
+  static const sxg_cert_t cert = {
+    .certificate = NULL,
+    .ocsp_response = NULL,
+    .sct_list = {NULL, 0, 0},
+  };
+  return cert;
+}
+
+sxg_cert_chain_t sxg_empty_cert_chain() {
+  static const sxg_cert_chain_t chain = {
+    .certs = NULL,
+    .size = 0,
+    .capacity = 0
+  };
+  return chain;
+}
+
+static void sxg_cert_release(sxg_cert_t* target) {
+  if (target == NULL) {
+    return;
+  }
+  X509_free(target->certificate);
+  OCSP_RESPONSE_free(target->ocsp_response);
+  sxg_buffer_release(&target->sct_list);
+  *target = sxg_empty_cert();
+}
+
+static bool ensure_free_capacity(size_t desired_margin,
+                                 sxg_cert_chain_t* target) {
+  return sxg_ensure_free_capacity_internal(
+      target->size, desired_margin, 8, sizeof(sxg_cert_chain_t), &target->capacity,
+      (void**)&target->certs);
+}
+
+void sxg_cert_chain_release(sxg_cert_chain_t* target) {
+  for (size_t i = 0; i < target->size; ++i) {
+    sxg_cert_release(&target->certs[i]);
+  }
+  OPENSSL_free(target->certs);
+  *target = sxg_empty_cert_chain();
+}
+
+static AUTHORITY_INFO_ACCESS* extract_aia_info(X509_EXTENSION* ext) {
+  const ASN1_OBJECT* const kAiaOid =
+      OBJ_txt2obj("1.3.6.1.5.5.7.1.1", /*dont_search_names=*/1);
+  if (ext == NULL || OBJ_cmp(X509_EXTENSION_get_object(ext), kAiaOid) != 0) {
+    return NULL;
+  }
+  AUTHORITY_INFO_ACCESS* aia_info = X509V3_EXT_d2i(ext);
+  if (aia_info == NULL) {
+    return NULL;
+  }
+  return aia_info;
+}
+
+bool sxg_extract_ocsp_url(X509* cert, sxg_buffer_t* dst) {
+  sxg_buffer_release(dst);
+  if (cert == NULL) {
+    return false;
+  }
+  const int extensions = X509_get_ext_count(cert);
+  for (int i = 0; i < extensions; ++i) {
+    AUTHORITY_INFO_ACCESS* const aia_info =
+        extract_aia_info(X509_get_ext(cert, i));
+    if (aia_info == NULL) {
+      continue;
+    }
+    for (int i = 0; i < sk_ACCESS_DESCRIPTION_num(aia_info); ++i) {
+      ACCESS_DESCRIPTION* ad = sk_ACCESS_DESCRIPTION_value(aia_info, i);
+      if (ad == NULL) {
+        continue;
+      }
+      int ad_nid = OBJ_obj2nid(ad->method);
+      switch (ad_nid) {
+        case NID_ad_OCSP: {
+          if (ad->location == NULL || ad->location->type != GEN_URI) {
+            continue;
+          }
+          ASN1_IA5STRING* uri = ad->location->d.uniformResourceIdentifier;
+          if (uri == NULL || uri->data == NULL || uri->length == 0) {
+            continue;
+          }
+          const bool success = sxg_write_bytes(uri->data, uri->length, dst) &&
+                         sxg_write_byte('\0', dst);
+          AUTHORITY_INFO_ACCESS_free(aia_info);
+          return success;
+        }
+        default:
+          break;
+      }
+    }
+    AUTHORITY_INFO_ACCESS_free(aia_info);
+  }
+  return false;
+}
+
+static bool wait_fd(int fd, bool read, bool write) {
```

Comment:
I'd suggest `epoll` instead of `select`.

Reply:
I priority on portability and readability rather than getting performance on switching `epoll` and `kqueue` with `ifdef` tricks.